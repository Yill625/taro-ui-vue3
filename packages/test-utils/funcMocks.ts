import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks