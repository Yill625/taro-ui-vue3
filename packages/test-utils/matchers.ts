import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers