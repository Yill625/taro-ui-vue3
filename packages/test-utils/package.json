{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}