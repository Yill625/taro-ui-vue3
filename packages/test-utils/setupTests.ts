import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()