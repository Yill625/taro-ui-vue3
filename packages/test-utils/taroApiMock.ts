import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
