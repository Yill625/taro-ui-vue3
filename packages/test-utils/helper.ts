import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}