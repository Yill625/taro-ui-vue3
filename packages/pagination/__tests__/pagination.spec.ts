import { mountFactory, Slots } from '@taro-ui-vue3/test-utils/helper'
import { ref } from '@vue/runtime-core'
import AtPagination from '../index'

const factory = (values = {}, slots: Slots = { default: [] }) => {
  return mountFactory(AtPagination, {}, values, slots)
}

describe('AtPagination', () => {
  it('should render prop -- current', () => {
    const wrapper = factory({ current: 2 })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop -- total', () => {
    const wrapper = factory({ total: 100 })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop -- pageSize', () => {
    const wrapper = factory({ pageSize: 40 })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop -- icon', () => {
    const wrapper = factory({ icon: true })
    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('AtPagination Behavior', () => {
  it('should trigger event onPageChange - prev', async () => {
    const current = ref(2)
    const onPageChange = jest.fn().mockImplementation((val) => {
      current.value = val.current
    })
    const wrapper = factory({
      total: 100,
      current: current.value,
      onPageChange,
    })
    expect(wrapper.element).toMatchSnapshot()
    await wrapper.find('.at-pagination .at-pagination__btn-prev .at-button').trigger('tap')
    expect(onPageChange).toBeCalled()
    expect(current.value).toBe(1)
  })

  it('should trigger event onPageChange - next', async () => {
    const current = ref(2)
    const onPageChange = jest.fn().mockImplementation((val) => {
      current.value = val.current
    })
    const wrapper = factory({
      total: 100,
      current: current.value,
      onPageChange,
    })
    expect(wrapper.element).toMatchSnapshot()
    await wrapper.find('.at-pagination .at-pagination__btn-next .at-button').trigger('tap')
    expect(onPageChange).toBeCalled()
    expect(current.value).toBe(3)
  })

  it('should not trigger event onPageChange when disabled', async () => {
    const onPageChange = jest.fn()
    const wrapper = factory({
      total: 20,
      onPageChange,
    })
    expect(wrapper.element).toMatchSnapshot()
    await wrapper.find('.at-pagination .at-pagination__btn-prev .at-button').trigger('tap')
    await wrapper.find('.at-pagination .at-pagination__btn-next .at-button').trigger('tap')
    expect(onPageChange).not.toBeCalled()
  })

  it('onPageChange should return params {type, current} - prev', async () => {
    const onPageChange = jest.fn()
    const wrapper = factory({
      current: 2,
      total: 100,
      onPageChange,
    })
    await wrapper.find('.at-pagination .at-pagination__btn-prev .at-button').trigger('tap')
    expect(onPageChange.mock.calls[0][0].type).toEqual('prev')
    expect(onPageChange.mock.calls[0][0].current).toEqual(1)
  })

  it('onPageChange should return params {type, current} - next', async () => {
    const onPageChange = jest.fn()
    const wrapper = factory({
      total: 100,
      onPageChange,
    })
    await wrapper.find('.at-pagination .at-pagination__btn-next .at-button').trigger('tap')
    expect(onPageChange.mock.calls[0][0].type).toEqual('next')
    expect(onPageChange.mock.calls[0][0].current).toEqual(2)
  })
})
