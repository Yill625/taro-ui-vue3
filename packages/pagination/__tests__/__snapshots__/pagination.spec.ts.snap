// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtPagination Behavior should not trigger event onPageChange when disabled 1`] = `
<view
  class="at-pagination"
>
  <view
    class="at-pagination__btn-prev"
  >
    <!---->
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        上一页
      </view>
    </view>
  </view>
  <view
    class="at-pagination__number"
  >
    <text
      class="at-pagination__number-current"
    >
      1
    </text>
    <text>
      /1
    </text>
  </view>
  <view
    class="at-pagination__btn-next"
  >
    <!---->
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        下一页
      </view>
    </view>
  </view>
</view>
`;

exports[`AtPagination Behavior should trigger event onPageChange - next 1`] = `
<view
  class="at-pagination"
>
  <view
    class="at-pagination__btn-prev"
  >
    <!---->
    <view
      class="at-button--small at-button"
    >
      <button
        class="at-button__wxbutton"
        lang="en"
      />
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        上一页
      </view>
    </view>
  </view>
  <view
    class="at-pagination__number"
  >
    <text
      class="at-pagination__number-current"
    >
      2
    </text>
    <text>
      /5
    </text>
  </view>
  <view
    class="at-pagination__btn-next"
  >
    <!---->
    <view
      class="at-button--small at-button"
    >
      <button
        class="at-button__wxbutton"
        lang="en"
      />
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        下一页
      </view>
    </view>
  </view>
</view>
`;

exports[`AtPagination Behavior should trigger event onPageChange - prev 1`] = `
<view
  class="at-pagination"
>
  <view
    class="at-pagination__btn-prev"
  >
    <!---->
    <view
      class="at-button--small at-button"
    >
      <button
        class="at-button__wxbutton"
        lang="en"
      />
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        上一页
      </view>
    </view>
  </view>
  <view
    class="at-pagination__number"
  >
    <text
      class="at-pagination__number-current"
    >
      2
    </text>
    <text>
      /5
    </text>
  </view>
  <view
    class="at-pagination__btn-next"
  >
    <!---->
    <view
      class="at-button--small at-button"
    >
      <button
        class="at-button__wxbutton"
        lang="en"
      />
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        下一页
      </view>
    </view>
  </view>
</view>
`;

exports[`AtPagination should render prop -- current 1`] = `
<view
  class="at-pagination"
>
  <view
    class="at-pagination__btn-prev"
  >
    <!---->
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        上一页
      </view>
    </view>
  </view>
  <view
    class="at-pagination__number"
  >
    <text
      class="at-pagination__number-current"
    >
      2
    </text>
    <text>
      /1
    </text>
  </view>
  <view
    class="at-pagination__btn-next"
  >
    <!---->
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        下一页
      </view>
    </view>
  </view>
</view>
`;

exports[`AtPagination should render prop -- icon 1`] = `
<view
  class="at-pagination at-pagination--icon"
>
  <view
    class="at-pagination__btn-prev"
  >
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        <text
          class="at-icon at-icon-chevron-left"
        />
      </view>
    </view>
    <!---->
  </view>
  <view
    class="at-pagination__number"
  >
    <text
      class="at-pagination__number-current"
    >
      1
    </text>
    <text>
      /1
    </text>
  </view>
  <view
    class="at-pagination__btn-next"
  >
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        <text
          class="at-icon at-icon-chevron-right"
        />
      </view>
    </view>
    <!---->
  </view>
</view>
`;

exports[`AtPagination should render prop -- pageSize 1`] = `
<view
  class="at-pagination"
>
  <view
    class="at-pagination__btn-prev"
  >
    <!---->
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        上一页
      </view>
    </view>
  </view>
  <view
    class="at-pagination__number"
  >
    <text
      class="at-pagination__number-current"
    >
      1
    </text>
    <text>
      /1
    </text>
  </view>
  <view
    class="at-pagination__btn-next"
  >
    <!---->
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        下一页
      </view>
    </view>
  </view>
</view>
`;

exports[`AtPagination should render prop -- total 1`] = `
<view
  class="at-pagination"
>
  <view
    class="at-pagination__btn-prev"
  >
    <!---->
    <view
      class="at-button--small at-button--disabled at-button"
    >
      <!---->
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        上一页
      </view>
    </view>
  </view>
  <view
    class="at-pagination__number"
  >
    <text
      class="at-pagination__number-current"
    >
      1
    </text>
    <text>
      /5
    </text>
  </view>
  <view
    class="at-pagination__btn-next"
  >
    <!---->
    <view
      class="at-button--small at-button"
    >
      <button
        class="at-button__wxbutton"
        lang="en"
      />
      <!---->
      <!---->
      <!---->
      <view
        class="at-button__text"
      >
        下一页
      </view>
    </view>
  </view>
</view>
`;
